guava/src/com/google/common/io/ReaderInputStream.java

```diff
@@ -185,6 +185,26 @@ public int read(byte[] b, int off, int len) throws IOException {
     }
   }
 
+  @Override
+  public int available() throws IOException {
+    if (byteBuffer.hasRemaining()) {
+      // if we have remaining chars in the buffer, return their count
+      return byteBuffer.remaining();
+    } else if (charBuffer.hasRemaining()) {
+      // if not, try to make quick progress by encoding available
+      // chars that have been read from the users reader
+      encoder.encode(charBuffer, byteBuffer, endOfInput);
+      return available();
+    } else if (reader.ready() && !endOfInput) {
```

Comment:
There has to be another branch checking `endOfInput && !doneFlushing` calling `encoder.flush(byteBuffer)` then (and checking for `OVERFLOW`, increasing the buffer, or throwing an exception is the result indicates an error).

Otherwise it might be possible that previously in the encode loop `encoder.flush(byteBuffer)` returned `OVERFLOW` and therefore `doneFlushing==false` and there are still pending bytes.

For example:
```java
// Requires Java 11+

int byteBufferSize = 8192; // Buffer size used by CharSource.AsByteSource.openStream()

// Charset whose CharsetEncoder.flush(...) puts bytes
Charset charset = Charset.forName("Cp930");

// \u0391: Causes CharsetEncoder.flush(...) to put additional bytes
String s = "a".repeat(byteBufferSize - 3) + "\u0391";
InputStream in = CharSource.wrap(s).asByteSource(charset).openStream();
in.read(new byte[byteBufferSize]);

// Should return 1
in.available();
```